price-feeder/oracle/provider/mexc.go

```diff
@@ -0,0 +1,504 @@
+		mexcPair := currencyPairToMexcPair(cp)
+		if tickerResp.Symbol[mexcPair].LastPrice != 0 {
+			p.setTickerPair(
+				mexcPair,
+				tickerResp.Symbol[mexcPair],
+			)
+			telemetry.IncrCounter(
+				1,
+				"websocket",
+				"message",
+				"type",
+				"ticker",
+				"provider",
+				string(ProviderMexc),
+			)
+			return
+		}
+	}
+
+	candleErr = json.Unmarshal(bz, &candleResp)
+	if candleResp.Metadata.Close != 0 {
+		p.setCandlePair(candleResp)
+		telemetry.IncrCounter(
+			1,
+			"websocket",
+			"message",
+			"type",
+			"candle",
+			"provider",
+			string(ProviderMexc),
+		)
+		return
+	}
+
+	if tickerErr != nil || candleErr != nil {
+		p.logger.Error().
+			Int("length", len(bz)).
+			AnErr("ticker", tickerErr).
+			AnErr("candle", candleErr).
+			Msg("mexc: Error on receive message")
+	}
+}
+
+func (p *MexcProvider) setTickerPair(symbol string, ticker MexcTickerData) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	price, err := sdk.NewDecFromStr(strconv.FormatFloat(ticker.LastPrice, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse ticker price")
+	}
+	volume, err := sdk.NewDecFromStr(strconv.FormatFloat(ticker.Volume, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse ticker volume")
+	}
+
+	p.tickers[symbol] = types.TickerPrice{
+		Price:  price,
+		Volume: volume,
+	}
+}
+
+func (p *MexcProvider) setCandlePair(candleResp MexcCandleResponse) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	close, err := sdk.NewDecFromStr(strconv.FormatFloat(candleResp.Metadata.Close, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse candle close")
+	}
+	volume, err := sdk.NewDecFromStr(strconv.FormatFloat(candleResp.Metadata.Volume, 'f', 5, 64))
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("mexc: failed to parse candle volume")
+	}
+	candle := types.CandlePrice{
+		Price:  close,
+		Volume: volume,
+		// convert seconds -> milli
+		TimeStamp: SecondsToMilli(candleResp.Metadata.TimeStamp),
+	}
+
+	staleTime := PastUnixTime(providerCandlePeriod)
+	candleList := []types.CandlePrice{}
+	candleList = append(candleList, candle)
+
+	for _, c := range p.candles[candleResp.Symbol] {
+		if staleTime < c.TimeStamp {
+			candleList = append(candleList, c)
+		}
+	}
+
+	p.candles[candleResp.Symbol] = candleList
+}
+
+func (p *MexcProvider) handleWebSocketMsgs(ctx context.Context) {
+	reconnectTicker := time.NewTicker(defaultMaxConnectionTime)
+	defer reconnectTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message.
+				p.logger.Err(err).Msg("mexc: could not read message")
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-reconnectTicker.C:
+			if err := p.disconnect(); err != nil {
```

Comment:
we are doing `reconnect`, so why do we need to do `disconnect`? Can we add `p.wsClient.Close()` to `reconnect`?